Sources/TuistCore/MetadataProviders/LibraryMetadataProvider.swift

```diff
@@ -2,19 +2,82 @@ import Foundation
 import TSCBasic
 import TuistSupport
 
-protocol LibraryMetadataProviding: PrecompiledMetadataProviding {
-    /// Returns the product for the given library.
-    /// - Parameter library: Library instance.
-    func product(library: LibraryNode) throws -> Product
+// MARK: - Metadata
+
+public struct LibraryMetadata: Equatable {
+    public var path: AbsolutePath
+    public var publicHeaders: AbsolutePath
+    public var swiftModuleMap: AbsolutePath?
+    public var architectures: [BinaryArchitecture]
+    public var linking: BinaryLinking
+}
+
+// MARK: - Provider
+
+public protocol LibraryMetadataProviding: PrecompiledMetadataProviding {
+    /// Loads all the metadata associated with a library (.a / .dylib) at the specified path
+    /// - Note: This performs various shell calls and disk operations
+    func loadMetadata(at path: AbsolutePath,
+                      publicHeaders: AbsolutePath,
+                      swiftModuleMap: AbsolutePath?) throws -> LibraryMetadata
+}
+
+// MARK: - Default Implementation
+
+public final class LibraryMetadataProvider: PrecompiledMetadataProvider, LibraryMetadataProviding {
+    public func loadMetadata(at path: AbsolutePath,
+                             publicHeaders: AbsolutePath,
+                             swiftModuleMap: AbsolutePath?) throws -> LibraryMetadata
+    {
+        let fileHandler = FileHandler.shared
+        guard fileHandler.exists(path) else {
+            throw LibraryMetadataProviderError.libraryNotFound(path)
+        }
+        guard fileHandler.exists(publicHeaders) else {
+            throw LibraryMetadataProviderError.publicHeadersNotFound(libraryPath: path, headersPath: publicHeaders)
+        }
+        if let swiftModuleMap = swiftModuleMap {
+            guard fileHandler.exists(swiftModuleMap) else {
+                throw LibraryMetadataProviderError.swiftModuleMapNotFound(libraryPath: path, moduleMapPath: swiftModuleMap)
+            }
+        }
+
+        let architectures = try self.architectures(binaryPath: path)
+        let linking = try self.linking(binaryPath: path)
+        return LibraryMetadata(
+            path: path,
+            publicHeaders: publicHeaders,
+            swiftModuleMap: swiftModuleMap,
+            architectures: architectures,
+            linking: linking
+        )
+    }
 }
 
-final class LibraryMetadataProvider: PrecompiledMetadataProvider, LibraryMetadataProviding {
-    func product(library: LibraryNode) throws -> Product {
-        switch try linking(binaryPath: library.path) {
-        case .dynamic:
-            return .dynamicLibrary
-        case .static:
-            return .staticLibrary
+// MARK: - Error
+
+enum LibraryMetadataProviderError: FatalError, Equatable {
```

Comment:
Like I mentioned in another comment, I'd move this one up to the top of the file.